### DISCUSSION TITLE
v2) support window close event control

### DISCUSSION THREAD

**Discussion by ironpark** [wailsapp/wails#977] [Ideas]
# Proposal
Name: OnCloseBefore or OnBeforeClose

Type: func(ctx context.Context) (prevent bool)

This callback is Called when the window close button is pressed, just before the application close. It provides the application context and if the return value is true , ignores the close event and does not close the application.

## Why
If we can control the close event fires or not
"There is work in progress. Are you sure you want to quit?" or "Are you sure you want to exit?" UX scenarios that prevent user mistakes, such as popping up a dialog like this, become possible.

## Example

```go
 func OnCloseBefore(ctx context.Context) (prevent bool){
    selection, _ := runtime.MessageDialog(b.ctx, runtime.MessageDialogOptions{
        Type:              runtime.WarningDialog,
        Title:               "There is work in progress!",
        Message:       "Are you sure you want to exit?",
        Buttons:       []string{"Yes", "No"},
        DefaultButton: "Yes",
        CancelButton:  "No",
    })
    return selection == "No"
 }
```

**Comment by leaanthony** [comment]
What do you think should happen if hide on close is set? Do you think it should trigger anyway?

**Reply by bahadirdogru** [reply]
Like electron we should be able to handle resize, close, hide, shown or others. For example if user close window i navigate main menu and if user close there i ask are you sure then i will close app. Or multiple window i open different pages and just show/hide to interact.
